cmd/ephemeral-installer-register-cluster-and-infraenv/main.go

```diff
@@ -0,0 +1,248 @@
+and the agent. The image is intend to be used by users to who are not currently
+supported by the IPI installer. It is also targeted for disconnected
+on-premises use.
+
+The client performs three main tasks:
+1. Reads in ZTP manifests containing information about cluster 0. The manifests
+   are either provided by users or generated by the ISO generation tool using
+   source files like install-config.yaml. The manifests are bind mounted
+   to the container running the client.
+2. Registers a cluster in assisted service using the REST-API.
+3. Registers a infraenv using the REST-API. The infraenv contains a
+   reference to the created cluster and contains any nmstate configs.
+
+A kube-api is not available in the ephemeral-installer environment. This
+is why the ZTP manifests are not applied directly to kubernetes. Instead,
+the manifests are translated into create cluster and infraenv params, and
+then posted to the REST-API to create those resources.
+
+This client is needed to register the cluster and infraenv to minimize
+user interaction and to enable automation use cases.
+
+A service running on the image, waits for the infraenv to be registered
+to assisted-service. Once registered, an infraenv id is available and
+the agent is started. The agent registers the host to assisted service
+allowing the installation to continue.
+
+A separate service waits for host validation to complete and initiates
+the cluster installation when the cluster status becomes ready.
+
+These two services are embedded in the image generated by the ephemeral
+installer tooling and are not part of this client.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/kelseyhightower/envconfig"
+	hiveext "github.com/openshift/assisted-service/api/hiveextension/v1beta1"
+	aiv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
+	"github.com/openshift/assisted-service/client"
+	"github.com/openshift/assisted-service/client/installer"
+	"github.com/openshift/assisted-service/internal/controller/controllers"
+	"github.com/openshift/assisted-service/models"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+)
+
+var Options struct {
+	ClusterDeploymentFile   string `envconfig:"CLUSTER_DEPLOYMENT_FILE" default:"/manifests/cluster-deployment.yaml"`
+	AgentClusterInstallFile string `envconfig:"AGENT_CLUSTER_INSTALL_FILE" default:"/manifests/agent-cluster-install.yaml"`
+	InfraEnvFile            string `envconfig:"INFRA_ENV_FILE" default:"/manifests/infraenv.yaml"`
+	PullSecretFile          string `envconfig:"PULL_SECRET_FILE" default:"/manifests/pull-secret.yaml"`
+	ClusterImageSetFile     string `envconfig:"CLUSTER_IMAGE_SET_FILE" default:"/manifests/cluster-image-set.yaml"`
+	NMStateConfigFile       string `envconfig:"NMSTATE_CONFIG_FILE" default:"/manifests/nmstateconfig.yaml"`
```

Comment:
Is that expected to be a single file?

Reply:
It is a single file right now. We will likely support multiple NMStateConfig files soon.

Reply:
So lets change it to a list?

Reply:
Correction, it will be a single file containing multiple yaml definitions. @bfournie has a PR in the fleeting repo to add multiple nmstateconfigs: https://github.com/openshift-agent-team/fleeting/pull/25. We will need to move his changes over here too.